File: OCR-Translate/Services/AnalysisService.swift
import Foundation

enum AnalysisError: LocalizedError {
    case emptyText
    case apiError(String)
    case parseError(String)

    var errorDescription: String? {
        switch self {
        case .emptyText: "没有可分析的文本"
        case .apiError(let msg): "分析出错: \(msg)"
        case .parseError(let msg): "解析分析结果出错: \(msg)"
        }
    }
}

final class AnalysisService {
    private let deepSeek = DeepSeekService()

    // MARK: - Prompts

    private func systemPrompt(for mode: AnalysisMode) -> String {
        switch mode {
        case .proofread:
            return """
            你是一位专业的日语校对专家。用户会给你一段日文文本（OCR 识别结果，没有换行和空格）。
            请先按句号「。」感叹号「！」问号「？」断句，然后对每个句子精准拆解语法成分。
            你需要精确标注每个成分对应的原文片段，帮助用户核对翻译是否准确。

            {
              "sentences": [
                {
                  "original": "完整的原句",
                  "components": [
                    {"label": "主语", "text": "原文片段", "explanation": "指代/施事者说明"},
                    {"label": "谓语", "text": "原文片段", "explanation": "时态/语态/敬体说明"},
                    {"label": "宾语", "text": "原文片段", "explanation": "受事对象"},
                    {"label": "定语", "text": "原文片段", "explanation": "修饰对象"},
                    {"label": "状语", "text": "原文片段", "explanation": "时间/地点/方式"},
                    {"label": "补语", "text": "原文片段", "explanation": "补充说明"}
                  ]
                }
              ]
            }

            要求：
            - 按句号「。」感叹号「！」问号「？」断句，逐句分析
            - 每一句必须包含主语和谓语，其他成分存在则标注，不存在则省略
            - explanation 用简洁中文说明该成分在句中的作用
            - 不需要语法点和词汇注解
            - 只输出 JSON，不要有任何其他文字
            """
        case .study:
            return """
            你是一位专业的日语教师，擅长教学分析。用户会给你一段日文文本（OCR 识别结果，没有换行和空格）。
            请按以下顺序分析：

            1. 按句号「。」感叹号「！」问号「？」断句
            2. 对每句拆解语法成分（主语·谓语·宾语·定语·状语·补语），标注原文片段和说明
            3. 列出关键语法点，附简要解释和 1 个例句
            4. 标注所有 JLPT N2~N1 级别词汇，以及值得学习的常用表达

            {
              "sentences": [
                {
                  "original": "完整的原句",
                  "components": [
                    {"label": "主语", "text": "原文片段", "explanation": "简要说明"},
                    {"label": "谓语", "text": "原文片段", "explanation": "简要说明"}
                  ],
                  "grammarPoints": ["语法格式：简要解释。例句：简短例句。"],
                  "vocabulary": [
                    {"word": "単語", "reading": "たんご", "meaning": "单词", "partOfSpeech": "名词", "jlptLevel": "N2", "notes": "常见用法/易错提示"}
                  ]
                }
              ],
              "overallNotes": "整体学习建议"
            }

            要求：
            - 语法点必须附 1 个简短例句，解释使用场景
            - 词汇注音用平假名，释义用中文，notes 可包含常见搭配或易错点
            - 不存在的成分省略
            - 只输出 JSON
            """
        }
    }

    // MARK: - Parse

    func analyze(text: String, mode: AnalysisMode) async throws -> AnalysisResult {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw AnalysisError.emptyText
        }

        let prompt = systemPrompt(for: mode)
        let rawResponse: String
        do {
            rawResponse = try await deepSeek.chat(
                systemPrompt: prompt,
                userMessage: trimmed,
                temperature: 0.2,
                maxTokens: 4096
            )
        } catch {
            throw AnalysisError.apiError(error.localizedDescription)
        }

        return try parse(rawResponse, mode: mode)
    }

    private func parse(_ jsonString: String, mode: AnalysisMode) throws -> AnalysisResult {
        let cleaned = jsonString
            .replacingOccurrences(of: "```json", with: "")
            .replacingOccurrences(of: "```", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        guard let data = cleaned.data(using: .utf8) else {
            throw AnalysisError.parseError("无法编码响应文本")
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AnalysisError.parseError("JSON 解析失败")
        }

        let sentencesJSON = json["sentences"] as? [[String: Any]] ?? []
        let sentences: [SentenceAnalysis] = sentencesJSON.compactMap { sentenceJSON in
            let original = sentenceJSON["original"] as? String ?? ""

            let componentsJSON = sentenceJSON["components"] as? [[String: Any]] ?? []
            let components = componentsJSON.compactMap { comp -> SentenceComponent? in
                guard let label = comp["label"] as? String,
                      let text = comp["text"] as? String else { return nil }
                return SentenceComponent(
                    label: label,
                    text: text,
                    explanation: comp["explanation"] as? String
                )
            }

            let grammarPoints = sentenceJSON["grammarPoints"] as? [String] ?? []

            let vocabJSON = sentenceJSON["vocabulary"] as? [[String: Any]] ?? []
            let vocabulary = vocabJSON.compactMap { vocab -> VocabularyAnnotation? in
                guard let word = vocab["word"] as? String,
                      let reading = vocab["reading"] as? String,
                      let meaning = vocab["meaning"] as? String else { return nil }
                return VocabularyAnnotation(
                    word: word,
                    reading: reading,
                    meaning: meaning,
                    partOfSpeech: vocab["partOfSpeech"] as? String,
                    jlptLevel: vocab["jlptLevel"] as? String,
                    notes: vocab["notes"] as? String
                )
            }

            return SentenceAnalysis(
                originalSentence: original,
                components: components,
                grammarPoints: grammarPoints,
                vocabulary: vocabulary
            )
        }

        return AnalysisResult(
            mode: mode,
            sentences: sentences,
            overallNotes: json["overallNotes"] as? String
        )
    }
}

File: win/NGMproofread.Windows/Services/AnalysisService.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using NGMproofread.Windows.Models;

namespace NGMproofread.Windows.Services;

public class AnalysisService
{
    private readonly AIService _ai = new();

    public async Task<AnalysisResult> AnalyzeAsync(string text, AnalysisMode mode)
    {
        var systemPrompt = mode == AnalysisMode.Study ? StudyPrompt : ProofreadPrompt;
        var rawResponse = await _ai.ChatAsync(systemPrompt, text, temperature: 0.3, maxTokens: 4096);
        return ParseResponse(rawResponse, mode);
    }

    private static AnalysisResult ParseResponse(string raw, AnalysisMode mode)
    {
        try
        {
            // Strip markdown code fences
            var json = Regex.Replace(raw, @"```(?:json)?\s*", "");
            json = Regex.Replace(json, @"```\s*$", "");
            json = json.Trim();

            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            var sentences = new List<SentenceAnalysis>();

            var sentencesArray = root.TryGetProperty("sentences", out var sa) ? sa : root;
            foreach (var sElem in sentencesArray.EnumerateArray())
            {
                var orig = sElem.GetProperty("original").GetString() ?? "";
                var components = ParseComponents(sElem);
                var grammarPoints = sElem.TryGetProperty("grammar_points", out var gp)
                    ? gp.EnumerateArray().Select(g => g.GetString() ?? "").Where(s => !string.IsNullOrEmpty(s)).ToList()
                    : [];
                var vocabulary = sElem.TryGetProperty("vocabulary", out var v)
                    ? ParseVocabulary(v)
                    : [];

                sentences.Add(new SentenceAnalysis
                {
                    OriginalSentence = orig,
                    Components = components,
                    GrammarPoints = grammarPoints,
                    Vocabulary = vocabulary
                });
            }

            string? notes = root.TryGetProperty("overall_notes", out var on) ? on.GetString() : null;

            return new AnalysisResult { Mode = mode, Sentences = sentences, OverallNotes = notes };
        }
        catch (Exception ex)
        {
            return new AnalysisResult
            {
                Mode = mode,
                Sentences = [new SentenceAnalysis { OriginalSentence = raw }],
                OverallNotes = $"解析失败: {ex.Message}"
            };
        }
    }

    private static List<SentenceComponent> ParseComponents(JsonElement parent)
    {
        if (!parent.TryGetProperty("components", out var comps)) return [];

        var result = new List<SentenceComponent>();
        foreach (var c in comps.EnumerateArray())
        {
            result.Add(new SentenceComponent
            {
                Label = c.GetProperty("label").GetString() ?? "",
                Text = c.GetProperty("text").GetString() ?? "",
                Explanation = c.TryGetProperty("explanation", out var e) ? e.GetString() : null,
                Children = ParseComponents(c)
            });
        }
        return result;
    }

    private static List<VocabularyAnnotation> ParseVocabulary(JsonElement vocabElem)
    {
        var result = new List<VocabularyAnnotation>();
        foreach (var v in vocabElem.EnumerateArray())
        {
            result.Add(new VocabularyAnnotation
            {
                Word = v.GetProperty("word").GetString() ?? "",
                Reading = v.GetProperty("reading").GetString() ?? "",
                Meaning = v.GetProperty("meaning").GetString() ?? "",
                PartOfSpeech = v.TryGetProperty("part_of_speech", out var pos) ? pos.GetString() : null,
                JlptLevel = v.TryGetProperty("jlpt_level", out var jl) ? jl.GetString() : null,
                Notes = v.TryGetProperty("notes", out var n) ? n.GetString() : null
            });
        }
        return result;
    }

    private const string ProofreadPrompt = """
        你是一个日语语法分析专家。请分析以下日语句子的语法成分。

        以 JSON 格式返回，结构如下：
        {
          "sentences": [
            {
              "original": "原始日语句子",
              "components": [
                {
                  "label": "主语",
                  "text": "对应的日文文本",
                  "explanation": "简短解释",
                  "children": [
                    {
                      "label": "定语",
                      "text": "修饰语文本",
                      "explanation": "解释"
                    }
                  ]
                }
              ]
            }
          ]
        }

        成分标签使用中文（主语、谓语、宾语、定语、状语、补语等）。
        成分树可以嵌套，children 表示子成分。
        只返回 JSON，不要额外解释。
        """;

    private const string StudyPrompt = """
        你是一个日语教学专家。请对以下日文句子进行深度解析。

        以 JSON 格式返回，结构如下：
        {
          "sentences": [
            {
              "original": "原始日语句子",
              "components": [
                {
                  "label": "主语",
                  "text": "对应的日文文本",
                  "explanation": "简短语法解释",
                  "children": []
                }
              ],
              "grammar_points": ["语法点1：解释+例句", "语法点2：解释+例句"],
              "vocabulary": [
                {
                  "word": "单词（日文）",
                  "reading": "读音（平假名）",
                  "meaning": "中文释义",
                  "part_of_speech": "词性（如：名词、他动词、形容动词等）",
                  "jlpt_level": "N1-N5 或 null",
                  "notes": "用法说明，可空"
                }
              ]
            }
          ],
          "overall_notes": "整体学习建议（可空）"
        }

        要求：
        1. components 进行完整的句子成分拆解，使用中文标签
        2. grammar_points 提取重要语法点，附带解释和例句
        3. vocabulary 重点标注 N1-N2 级别的较难词汇
        4. overall_notes 给出学习建议
        只返回 JSON，不要额外解释。
        """;
}
